### PR TITLE
feat: nested CORS iframes, ignore controls, and closed shadow DOM

### PR DESCRIPTION
**Pull request by aryanku-dev**
## Summary
Brings percy-selenium-java to parity with the canonical Percy CORS iframe + closed shadow DOM feature set.

## Implemented
- Nested cross-origin iframe capture (depth-capped, cycle-guarded)
- `data-percy-ignore` attribute opt-out
- `ignoreIframeSelectors` option
- Post-switch URL re-check via `isUnsupportedIframeSrc`
- `PercyContextLostException` recovery merges `partialCapture`
- Closed shadow DOM via CDP (`exposeClosedShadowRoots`)
- Inlined Java helpers (`clampFrameDepth`, `normalizeIgnoreSelectors`, `resolveMaxFrameDepth`, `resolveIgnoreSelectors`)

## Skipped
- ElementInternals preflight (Feature 8): N/A — selenium-java has no before-page-load hook.
- @percy/sdk-utils bump (Feature 9): not applicable to Java; helpers inlined.

## Reference
Mirrored from percy/percy-nightwatch#869 (PER-7292-add-cors-iframe-support); CDP from percy/percy-playwright#609.

## Test plan
- [x] New `IframeFeatureTest` (11 tests) and existing `CacheTest` (3 tests) pass under `mvn test`.
- [ ] `SdkTest`'s integration tests require a local Firefox binary (`@BeforeAll` instantiates `FirefoxDriver`). They were not exercised in the sync environment because Firefox is not installed; this matches the pre-existing baseline on `master` and is not a regression introduced by this PR.
- [ ] Manual smoke: cross-origin iframes
- [ ] Manual smoke: closed shadow roots in Chrome

🤖 Generated with [Claude Code](https://claude.com/claude-code) via /percy-sdk-sync

### CODE REVIEW

**Comment by aryanku-dev** [comment]
## 🤖 stack:pr-review — Automated review (BrowserStack AI harness)

**Summary:** Adds nested CORS iframe capture (depth cap + cycle guard), `data-percy-ignore` / `ignoreIframeSelectors` skipping, post-switch URL re-validation, partial-capture recovery on lost frame context, and closed-shadow-DOM exposure via CDP for Chromium — plus 15 new mock-based unit tests.

| Priority | Category | Check | Status |
|---|---|---|---|
| High | Security | No hardcoded secrets | Pass |
| High | Security | Input validation/sanitization | Pass — selectors normalized, depth clamped, `Element.matches` per-selector try/catch, URLs parsed via `URI` |
| High | Security | AuthZ / IDOR / SQLi | N/A |
| High | Correctness | Logic & edge cases | Pass — cycle guard via `ancestorUrls`, null-safe `Objects.equals` origin compare, parent-frame restore with `defaultContent` fallback |
| High | Correctness | Explicit error handling | Pass — `PercyContextLostException` carries partial capture; `FatalIframeException` re-thrown; per-iframe failures debug-logged |
| High | Correctness | No race conditions | Pass — single-threaded; `DOM.disable` & script timeout released in `finally` |
| Medium | Testing | New code / error paths tested | Pass — 15 new tests in `IframeFeatureTest` |
| Medium | Testing | Existing tests pass | Pass — 55/56; sole error is Firefox-missing `SdkTest.testSetup` (env, CI green) |
| Medium | Performance | No unbounded fetch | Pass — recursion bounded by `maxFrameDepth` |
| Medium | Quality | Patterns / focused | Pass — mirrors canonical `percy-playwright-java` |
| Low | Quality | Names / comments / deps | Pass |

**Findings:** none. Notable improvement: `getSerializedDOM` now throws a clear `RuntimeException` on non-`Map` serialize output (caught & degraded gracefully by `snapshot()`) instead of NPE-ing — strictly better than `main`.

**Verification:** `mvn -o compile` clean; `IframeFeatureTest` 15/15 pass; full suite 55/56 (only the env-dependent Firefox setup fails).

### Overall: ✅ **Pass** — Approve. No code changes needed. ⚠️ **Human blocker:** the pending **Percy visual-diff approval** must be approved in the Percy UI.

<sub>Posted by the BrowserStack `stack:pr-review` harness.</sub>

File: src/main/java/io/percy/selenium/Percy.java [hunk 1]
```diff
@@ -737,94 +757,378 @@ private String getOrigin(String url) {
...
+            // contentDocument subtrees are skipped during collection so this
+            // remains scoped to the host element-level pairs visible from the
+            // current driver session. Non-fatal — wrapped in try/catch so a
+            // non-Chromium driver or restricted frame falls through silently.
+            // TODO(closed-shadow-cors): drive per-frame CDP via flat sessions
+            // (Target.setAutoAttach / sessionId) for deeper isolation when
+            // Selenium 4 BiDi support stabilises across versions.
+            try { exposeClosedShadowRoots(driver); } catch (Exception ignore) {}
+
+            Map<String, Object> iframeSnapshot = serializeCurrentFrame(options);
```
Comment (aryanku-dev) [inline src/main/java/io/percy/selenium/Percy.java:983]: **[High] Null iframeSnapshot silently poisons the CORS payload**

`serializeCurrentFrame` unchecked-casts the `PercyDOM.serialize` result and returns `null` when the script yields null (e.g. `PercyDOM` failed to load in a sandboxed CORS frame). That `null` is then placed straight into the collected entry below, producing a malformed `corsIframes` payload. The main-page path `getSerializedDOM` (~line 1063) explicitly throws on a non-Map result — this path lacks the equivalent guard.

**Suggestion:** After the call, if the snapshot is null/non-Map, log a warning and `return collected;` (skip this frame) rather than adding an entry with a null snapshot.

_Reviewer: stack:code-reviewer_

**Review by aryanku-dev — COMMENTED** [review]
Claude Code Review (automated) — 5 inline finding(s). Full report in the PR comment below. Verdict: Failed - see PR comment.

File: src/main/java/io/percy/selenium/Percy.java [hunk 2]
```diff
@@ -843,6 +1147,109 @@ Map<String, Object> getSerializedDOM(JavascriptExecutor jse, Set<Cookie> cookies
...
 
+    // Discover closed shadow roots via CDP and expose them on a window-bound
+    // WeakMap that PercyDOM.serialize reads to pierce closed shadow DOM. This is
+    // Chromium-only — wrapped in try/catch so other browsers (or a missing
+    // executeCdpCommand) fall through silently. Three CDP calls per pair:
+    // DOM.getDocument (depth=-1, pierce=true) to discover, then DOM.resolveNode
+    // for host + shadow, then Runtime.callFunctionOn to write the pair into
+    // the WeakMap on the page.
+    @SuppressWarnings("unchecked")
+    private void exposeClosedShadowRoots(WebDriver driver) {
```
Comment (aryanku-dev) [inline src/main/java/io/percy/selenium/Percy.java:1194]: **[Medium] CDP call bypasses the existing isCdpSupported guard**

`exposeClosedShadowRoots` issues `executeCdpCommand` directly and relies on an outer try/catch, while `changeWindowDimensionAndWait` (~line 1275) guards with `isCdpSupported(ChromeDriver)`. Use the same pattern for consistency.

**Suggestion:** `if (!(driver instanceof ChromeDriver)) return;` then `if (!isCdpSupported(chrome)) return;` before any CDP call.

_Reviewer: stack:code-reviewer_

File: src/main/java/io/percy/selenium/Percy.java [hunk 3]
```diff
@@ -737,94 +757,378 @@ private String getOrigin(String url) {
...
+    }
+
+    private List<String> resolveIgnoreSelectors(Map<String, Object> options) {
+        if (options != null && options.containsKey("ignoreIframeSelectors")) {
+            return normalizeIgnoreSelectors(options.get("ignoreIframeSelectors"));
+        }
+        if (cliConfig != null && cliConfig.has("snapshot") && !cliConfig.isNull("snapshot")) {
+            JSONObject snap = cliConfig.getJSONObject("snapshot");
+            if (snap.has("ignoreIframeSelectors") && !snap.isNull("ignoreIframeSelectors")) {
+                JSONArray arr = snap.optJSONArray("ignoreIframeSelectors");
```
Comment (aryanku-dev) [inline src/main/java/io/percy/selenium/Percy.java:804]: **[Medium] Single-string ignoreIframeSelectors CLI config is silently dropped**

`snap.optJSONArray("ignoreIframeSelectors")` returns null for a JSON string value (e.g. `"ignoreIframeSelectors": "iframe.ads"`), so a string-valued config is dropped — inconsistent with the options-map path that normalizes strings.

**Suggestion:** When `optJSONArray` is null, fall back to `optString` and run it through `normalizeIgnoreSelectors`.

_Reviewer: stack:code-reviewer_

File: src/main/java/io/percy/selenium/Percy.java [hunk 4]
```diff
@@ -737,94 +757,378 @@ private String getOrigin(String url) {
...
+
+    // True if the iframe element matches any of the user-provided ignore selectors.
+    // Selector matching is performed in-browser via Element.matches so any CSS
+    // selector the browser supports is valid; invalid selectors are tolerated.
+    private boolean iframeMatchesIgnoreSelector(WebElement iframe, List<String> selectors) {
+        if (selectors == null || selectors.isEmpty()) return false;
+        try {
+            JavascriptExecutor jse = (JavascriptExecutor) driver;
+            JSONArray sel = new JSONArray(selectors);
+            String script = "var el = arguments[0]; var selectors = " + sel.toString() + ";"
```
Comment (aryanku-dev) [inline src/main/java/io/percy/selenium/Percy.java]: **[Medium] Selector list interpolated into the script body**

The script is built by concatenating `sel.toString()` (a JSON array literal) into the JS source. A selector containing a quote or control character breaks the literal and throws a script error; the inner `el.matches` try/catch only handles invalid CSS.

**Suggestion:** Pass the selector list as an `executeScript` argument (`arguments[1]`) instead of interpolating it.

_Reviewer: stack:code-reviewer_

File: src/main/java/io/percy/selenium/Percy.java [hunk 5]
```diff
@@ -737,94 +757,378 @@ private String getOrigin(String url) {
...
+        String frameSrc = frameElement.getAttribute("src");
         String percyElementId = frameElement.getAttribute("data-percy-element-id");
-        log("processFrame: data-percy-element-id=\"" + percyElementId + "\" for src=\"" + finalFrameUrl + "\"", "debug");
+
+        List<Map<String, Object>> collected = new ArrayList<>();
+        if (depth > maxFrameDepth) {
+            log("Reached max iframe nesting depth (" + maxFrameDepth + "); stopping at " + frameSrc, "debug");
+            return collected;
+        }
+        if (ancestorUrls != null && frameSrc != null && ancestorUrls.contains(frameSrc)) {
```
Comment (aryanku-dev) [inline src/main/java/io/percy/selenium/Percy.java:932]: **[Medium] Cycle guard can miss a relative-vs-absolute src cycle**

The guard checks `ancestorUrls.contains(frameSrc)` on the raw `src`. The ancestor set is seeded with both raw `frameSrc` and absolute `reportedUrl`, so identical recurring `src` strings are caught — but a cycle expressed as a relative src at one level and an absolute URL at another can slip through. Also untested.

**Suggestion:** Resolve `frameSrc` to absolute (against the parent URL) before seeding/checking the ancestor set, and add a cycle-guard test.

_Reviewer: stack:code-reviewer_

**Comment by aryanku-dev** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-selenium-java/pull/312  •  **Head:** 4b7835d  •  **Reviewers:** stack:code-reviewer

## Summary
Adds cross-origin (CORS) iframe capture and closed shadow DOM exposure to the Percy Selenium Java SDK: recursive iframe traversal with a depth cap and cycle guard, `data-percy-ignore` / `ignoreIframeSelectors` honouring, partial-capture recovery on lost frame context via `PercyContextLostException`, closed shadow DOM exposure via CDP `DOM.getDocument(pierce=true)` for Chromium, plus null-safety hardening of the responsive-config read and supporting tests.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | None introduced. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface in this change. |
| High | Security | Input validation and sanitization | Pass | Iframe `src`/scheme validated; see #6 (selector interpolation, defensive). |
| High | Security | No IDOR — resource ownership validated | N/A | Not applicable. |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL. |
| High | Correctness | Logic is correct, handles edge cases | Fail | #2: null `iframeSnapshot` enters the CORS entry, unlike the guarded main-page path. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Frame errors logged; partial capture propagated via `PercyContextLostException`. |
| High | Correctness | No race conditions or concurrency issues | Pass | Sequential recursion; no shared mutable state across threads. |
| Medium | Testing | New code has corresponding tests | Pass | Extensive new tests; gaps noted below (cycle guard, depth-cap recursion). |
| Medium | Testing | Error paths and edge cases tested | Pass | NPE/clamp/bad-pair/context-lost paths covered. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Existing tests retargeted, not weakened. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Recursion bounded by depth cap. |
| Medium | Performance | Long-running tasks use background jobs | N/A | Not applicable. |
| Medium | Quality | Follows existing codebase patterns | Fail | #4: bypasses the existing `isCdpSupported` guard used elsewhere. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scoped to CORS iframes + shadow DOM (+ related null-safety). |
| Low | Quality | Meaningful names, no dead code | Fail | #8: pre-existing unused imports in a touched file. |
| Low | Quality | Comments explain why, not what | Pass | Comments are explanatory; #12 minor inaccuracy. |
| Low | Quality | No unnecessary dependencies added | Pass | Only a `packageManager` field added (see #9). |

## Findings

### High

- **File:** `src/main/java/io/percy/selenium/Percy.java:956` (helper at `:874`)
- **Severity:** High
- **Reviewer:** stack:code-reviewer (confirmed)
- **Issue:** `serializeCurrentFrame` does an unchecked cast of `jse.executeScript("return PercyDOM.serialize(...)")` and returns `null` when the script yields null (e.g. `PercyDOM` failed to load in a sandboxed CORS frame). The call site puts that `null` straight into the collected entry (`entry.put("iframeSnapshot", iframeSnapshot)`), producing a poisoned `corsIframes` payload. The main-page path `getSerializedDOM` (lines 1063-1066) explicitly throws on a non-Map result — this CORS path lacks the equivalent guard, an asymmetry that surfaces a malformed snapshot silently to the CLI.
- **Suggestion:** After the call, guard for null/non-Map: log a warning and `return collected;` (skip the frame) rather than adding an entry with a null snapshot — mirroring the main-page guard.

### Medium

- **File:** `src/main/java/io/percy/selenium/Percy.java:1158`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `exposeClosedShadowRoots` issues `executeCdpCommand` directly and relies on an outer try/catch, instead of the existing `isCdpSupported(ChromeDriver)` guard used by `changeWindowDimensionAndWait` (line 1275). Inconsistent with the established pattern.
- **Suggestion:** Gate with `if (!(driver instanceof ChromeDriver)) return;` then `if (!isCdpSupported(chrome)) return;` before any CDP call.

- **File:** `src/main/java/io/percy/selenium/Percy.java:800`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `resolveIgnoreSelectors` reads `snap.optJSONArray("ignoreIframeSelectors")`, which returns `null` for a JSON *string* value (e.g. `"ignoreIframeSelectors": "iframe.ads"`). The string config is then silently dropped, inconsistent with the options-map path that normalizes strings.
- **Suggestion:** When `optJSONArray` is null, fall back to `optString` and run it through `normalizeIgnoreSelectors`.

- **File:** `src/main/java/io/percy/selenium/Percy.java:819`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `iframeMatchesIgnoreSelector` builds the script by string-concatenating `sel.toString()` (a JSON array literal) into the JS body. A selector containing a quote or JS control character would break the literal and throw a script error (the inner `el.matches` try/catch only handles invalid CSS, not malformed JS).
- **Suggestion:** Pass the selector list as an `executeScript` argument (`arguments[1]`) instead of interpolating it into the script body.

- **File:** `src/main/java/io/percy/selenium/Percy.java:916`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer (severity adjusted by orchestrator)
- **Issue:** The cycle guard checks `ancestorUrls.contains(frameSrc)` using the raw `src` attribute. The ancestor set is seeded with *both* the raw `frameSrc` and the absolute `reportedUrl` (lines 979-980), so an identical recurring `src` string *is* caught — but a cycle expressed as a relative `src` at one level and an absolute URL at another can slip through. Narrower than "fails for the common case," but a real edge-case gap, and it is untested.
- **Suggestion:** Resolve `frameSrc` to an absolute URL (against the parent URL) before seeding and checking the ancestor set, so comparisons are consistently absolute. Add a cycle-guard test.

### Low

- **File:** `src/main/java/io/percy/selenium/Percy.java:32-33`
- **Severity:** Low — **Issue:** Unused imports `javax.swing.text.html.CSS` and `javax.xml.xpath.XPath` (pre-existing) live in a file this PR touches. **Suggestion:** Remove them.

- **File:** `package.json:11`
- **Severity:** Low — **Issue:** The `packageManager` SHA-512 digest is 126 chars (Corepack requires the full 128-char hex digest); Corepack may reject or silently fall back. **Suggestion:** Regenerate via `corepack prepare yarn@1.22.22 --activate`, or drop the field.

- **File:** `src/test/java/io/percy/selenium/IframeFeatureTest.java:1101`
- **Severity:** Low — **Issue:** A test reads `Percy.java` via the cwd-relative path `src/main/java/...`, which breaks when the runner's working directory isn't the repo root; it also asserts on source text rather than behaviour. **Suggestion:** Resolve via classpath/test-class location, or replace the source-text assertion with a behavioural one.

- **File:** `src/main/java/io/percy/selenium/Percy.java:1089` — **Severity:** Low — **Issue:** `ctx` map uses stringly-typed keys (`"options"`, `"maxFrameDepth"`, `"ignoreSelectors"`) with unchecked casts; a key typo fails at runtime. **Suggestion:** Replace with a small value object / record. (Style; non-blocking.)

- **File:** `src/test/java/io/percy/selenium/IframeFeatureTest.java` — **Severity:** Low — **Issue:** Comment in `getSerializedDomRecoversPartialCaptureOnContextLost` says "Five executeScript phases" but describes four. **Suggestion:** Fix the comment.

### Missing test coverage (non-gating)
- Cycle detection (the `ancestorUrls.contains(frameSrc)` guard) is not exercised by any test.
- Depth-cap enforcement via actual recursion (`processFrameTree` at `depth = maxFrameDepth + 1`) is not integration-tested; only the clamp helper is unit-tested.
- `resolveIgnoreSelectors` with a single-string CLI config value (finding #5) is untested.

---

**Verdict: FAIL** — one High-severity correctness issue (#2: null iframe snapshot silently poisons the CORS payload). The Medium findings (#4–#7) are worth folding in before merge; the Low items can follow up.

File: src/test/java/io/percy/selenium/IframeFeatureTest.java [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+
+    TargetLocator targetLocator = mock(TargetLocator.class);
+    when(mockedDriver.switchTo()).thenReturn(targetLocator);
+    when(targetLocator.frame(iframe)).thenReturn(mockedDriver);
+    when(targetLocator.defaultContent()).thenReturn(mockedDriver);
+    when(targetLocator.parentFrame()).thenReturn(mockedDriver);
+
+    when(((JavascriptExecutor) mockedDriver).executeScript(any(String.class)))
+      .thenReturn(null)
+      .thenReturn("https://cdn.other.com/frame");
```
Comment (aryanku-dev) [inline src/test/java/io/percy/selenium/IframeFeatureTest.java:292]: **[Medium] Brittle positional stub in the cycle-guard test**

This `.thenReturn(null).thenReturn("https://cdn.other.com/frame")` chain passes today, but if a future change inserts another `executeScript` call between the dom.js injection and `readCurrentFrameUrl()`, the URL stub fires at the wrong call site — `postSwitchUrl` becomes null, the cycle check is skipped, and the test stays green while no longer validating the cycle guard.

**Suggestion:** Use a script-content-keyed `thenAnswer` (match `return document.URL`), as the null-serialize and context-lost tests already do.

_Reviewer: stack:code-reviewer_

File: src/test/java/io/percy/selenium/IframeFeatureTest.java [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+    ctx.put("options", new HashMap<String, Object>());
+    ctx.put("maxFrameDepth", 5);
+    ctx.put("ignoreSelectors", java.util.Collections.<String>emptyList());
+
+    @SuppressWarnings("unchecked")
+    List<Map<String, Object>> result = (List<Map<String, Object>>) invokePrivate(
+      percy, "processFrameTree",
+      new Class[]{WebElement.class, int.class, Set.class, Map.class},
+      iframe, 1, new HashSet<String>(), ctx);
+    assertTrue(result.isEmpty(), "Frame must be skipped when PercyDOM.serialize returns null");
```
Comment (aryanku-dev) [inline src/test/java/io/percy/selenium/IframeFeatureTest.java:269]: **[Low] Test doesn't verify the frame was stepped out of**

The test asserts the result is empty but doesn't confirm `parentFrame()` ran in the `finally`. Adding the verify would guard against a future change that bails before `switchedIn = true`.

**Suggestion:** Add `verify(targetLocator, times(1)).parentFrame();` after the assertion.

_Reviewer: stack:code-reviewer_

**Review by aryanku-dev — COMMENTED** [review]
Claude Code Review (automated) — 2 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by aryanku-dev** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-selenium-java/pull/312  •  **Head:** c7ffcb1  •  **Reviewers:** stack:code-reviewer

_Continues [the previous review](https://github.com/percy/percy-selenium-java/pull/312#issuecomment-4810175790) — changes since `4b7835d` (delta re-review)._

## Summary
Delta commit `c7ffcb1` addresses the previous review's findings on the CORS iframe + closed shadow DOM work: null-snapshot skip in `serializeCurrentFrame`/`processFrameTree`, an absolute-URL post-switch cycle check, an `isCdpSupported` guard in `exposeClosedShadowRoots`, a single-string `ignoreIframeSelectors` CLI fallback, and selector values passed as a script argument instead of being interpolated — plus three new unit tests. Unused imports removed.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | None. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface. |
| High | Security | Input validation and sanitization | Pass | Selector values no longer interpolated into the script body (resolved). |
| High | Security | No IDOR — resource ownership validated | N/A | — |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL. |
| High | Correctness | Logic is correct, handles edge cases | Pass | Prior High (null snapshot) resolved and verified. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Null path logs at `warn`; `finally` still steps out of the frame. |
| High | Correctness | No race conditions or concurrency issues | Pass | Sequential recursion. |
| Medium | Testing | New code has corresponding tests | Pass | Three new tests; one brittle stub (Finding 1). |
| Medium | Testing | Error paths and edge cases tested | Pass | Null-snapshot, resolved-URL cycle, single-string CLI now covered. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | IframeFeatureTest 18/18, CacheTest 3/3, PercyStepsTest 37/37 locally. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Depth-bounded recursion. |
| Medium | Performance | Long-running tasks use background jobs | N/A | — |
| Medium | Quality | Follows existing codebase patterns | Pass | `isCdpSupported` guard now matches the established pattern. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scoped to the review fixes. |
| Low | Quality | Meaningful names, no dead code | Pass | Unused imports removed; one pre-existing unreachable catch noted (Finding 4). |
| Low | Quality | Comments explain why, not what | Pass | New comments are rationale-focused. |
| Low | Quality | No unnecessary dependencies added | Pass | — |

## Findings

### Resolved since last review
- ~~`src/main/java/io/percy/selenium/Percy.java:956` High — Null iframeSnapshot from serializeCurrentFrame poisons CORS entry~~ **Resolved** — `instanceof Map` guard + null early-return in `processFrameTree`; verified.
- ~~`src/main/java/io/percy/selenium/Percy.java:1158` Medium — exposeClosedShadowRoots skips isCdpSupported guard~~ **Resolved** — `if (!isCdpSupported(chrome)) return;` added.
- ~~`src/main/java/io/percy/selenium/Percy.java:800` Medium — resolveIgnoreSelectors drops single-string CLI config value~~ **Resolved** — `optString` fallback added + test.
- ~~`src/main/java/io/percy/selenium/Percy.java:819` Medium — iframeMatchesIgnoreSelector interpolates selectors into script body~~ **Resolved** — selectors passed as `arguments[1]`.
- ~~`src/main/java/io/percy/selenium/Percy.java:916` Medium — Cycle guard can miss relative-vs-absolute src mismatch~~ **Resolved** — post-switch absolute-URL check added + test.
- ~~`package.json:11` Low — Truncated packageManager SHA-512 hash~~ **Resolved (false positive)** — hash is a valid 128-char digest; no change needed.

### Still open (non-gating)

- **File:** `src/test/java/io/percy/selenium/IframeFeatureTest.java:292`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `processFrameTreeSkipsCyclicFrameByResolvedUrl` uses a positional `.thenReturn(null).thenReturn("https://cdn.other.com/frame")` stub. It passes today, but if a future change inserts another `executeScript` call between the dom.js injection and `readCurrentFrameUrl()`, the URL stub fires at the wrong site and the test silently exercises a different path while staying green.
- **Suggestion:** Use a script-content-keyed `thenAnswer` (matching `return document.URL`), as the null-serialize and context-lost tests already do.

- **File:** `src/test/java/io/percy/selenium/IframeFeatureTest.java:269`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** `processFrameTreeSkipsWhenSerializeReturnsNull` asserts the result is empty but doesn't verify the frame was stepped out of. Adding `verify(targetLocator).parentFrame()` would guard against a future change that bails before `switchedIn = true`.
- **Suggestion:** Add `verify(targetLocator, times(1)).parentFrame();`.

- **File:** `src/test/java/io/percy/selenium/IframeFeatureTest.java:590`
- **Severity:** Low (carried forward)
- **Reviewer:** stack:code-reviewer
- **Issue:** `exposeClosedShadowRootsIsAttemptedInsideCorsFrame` reads `Percy.java` via the cwd-relative path `src/main/java/...`, which breaks when the runner's working directory isn't the repo root. This delta didn't touch that test, so it remains open.
- **Suggestion:** Resolve the path from the class/code-source location, or assert behaviour rather than source text.

- **File:** `src/main/java/io/percy/selenium/Percy.java:1190`
- **Severity:** Low (informational, pre-existing)
- **Reviewer:** stack:code-reviewer
- **Issue:** After `if (!(driver instanceof ChromeDriver)) return;`, the subsequent `try { chrome = (ChromeDriver) driver; } catch (ClassCastException e)` is unreachable. Not introduced by this delta; the new `isCdpSupported` line sits right beside it.
- **Suggestion:** Drop the redundant try/catch in a future cleanup.

---

**Verdict: PASS** — every gating finding from the previous review (1 High + 4 Mediums) is resolved and the fixes were independently verified as correct. The remaining items are non-gating test-quality nits and one pre-existing informational note.

**Comment by aryanku-dev** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-selenium-java/pull/312  •  **Head:** 80fadac  •  **Reviewers:** stack:code-reviewer

## Summary
Adds nested cross-origin (CORS) iframe capture and closed-shadow-DOM exposure (Chromium CDP). Replaces single-level `processFrame` with a recursive `processFrameTree` bounded by a clamped depth (default 3, cap 10), dual cycle guards (raw `src` + resolved `document.URL`), and a `PercyContextLostException` that salvages partial captures. Adds `data-percy-ignore` + `ignoreIframeSelectors` filtering, post-switch URL re-checks against the canonical 15-prefix scheme list, null-safe origin comparison, and a Chrome-only `exposeClosedShadowRoots`. CI is green across Java 8/11/17/21, Analyze, CodeQL, Semgrep. Depth/scheme alignment is canonical (percy/cli#2319).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets | Pass | No credentials/tokens; the `packageManager` hash is a yarn SRI, not a secret. |
| High | Security | Authn/authz enforced | N/A | Client-side DOM serialization SDK; no auth surface. |
| High | Security | Input validation | Pass | Selectors passed as JS script args (not interpolated); list trimmed/null-filtered; depth coerced+clamped; scheme list filters `javascript:`/`data:`/`file:`/`blob:` etc. |
| High | Security | No IDOR | N/A | No resource access by user-supplied id. |
| High | Security | No SQL injection | N/A | No SQL/datastore queries. |
| High | Correctness | Logic/edge cases | Pass | Depth clamp (0/neg→default, >cap→10), dual cycle guards, null/non-Map serialize → skip, null-origin via Objects.equals; covered by tests. |
| High | Correctness | Error handling | Pass | Per-iframe and per-pair try/catch; `PercyContextLostException` propagates partial capture; `finally` steps up one frame with defaultContent fallback; CDP non-fatal with `DOM.disable` cleanup in finally. |
| High | Correctness | No race conditions | Pass | Single-threaded driver interaction; recursion uses per-call ancestor copies; no shared mutable state. |
| Medium | Testing | New code tested | Pass | 15-test IframeFeatureTest covers clamp, normalize, resolve (option/CLI/scalar), ignore paths, post-switch skip, null-serialize skip, cyclic, partial capture, non-Chrome no-op, collector walk. |
| Medium | Testing | Error/edge cases tested | Pass | NPE/null-origin, clamp(0), unsupported-scheme post-switch, context-lost recovery, bad-pair tolerance, contentDocument skip exercised. |
| Medium | Testing | Existing tests pass | Pass | CI Test on Java 8/11/17/21 all green. |
| Medium | Performance | No N+1 / unbounded | Pass | Recursion bounded by depth cap (10) and cycle guards; `DOM.getDocument` called once per invocation. |
| Medium | Performance | Heavy work backgrounded | N/A | Synchronous SDK capture path. |
| Medium | Quality | Follows patterns | Pass | Reuses existing `isCdpSupported`/CDP gating, `log` levels, `getOrigin`, JSON config access patterns. |
| Medium | Quality | Focused change | Pass | Scoped to CORS-iframe + closed-shadow capture plus related defensive guards; stray-import removal is reasonable housekeeping. |
| Low | Quality | Names / no dead code | Pass | Clear names; removed unused `javax.swing`/`javax.xml` imports. |
| Low | Quality | Comments explain why | Pass | Comments justify depth=0 semantics, parent-vs-page origin compare, post-switch rationale, CDP gating, and a future-BiDi TODO. |
| Low | Quality | No unnecessary deps | Pass | No new runtime deps; package.json adds only a `packageManager` field; CDP uses existing ChromeDriver API. |

## Findings
No blocking issues found.

Non-blocking observations (informational):
- Depth default (3) / cap (10) and the 15 case-insensitive scheme prefixes align with canonical `@percy/sdk-utils` (percy/cli#2319); guarded by `clampFrameDepthZeroReturnsDocumentedDefault`.
- Closed-shadow exposure inside CORS frames is acknowledged best-effort with a documented `TODO(closed-shadow-cors)` for per-frame flat-session isolation once Selenium 4 BiDi stabilizes.

_Note: the PR's Percy "visual change needs review" status is a manual visual-approval gate in the Percy dashboard, separate from this code review._

---

**Verdict: PASS**

**Review by pranavz28 — APPROVED** [review]
